src/metabase/domain_entities/converters.cljs

```diff
@@ -1,73 +1,95 @@
 
-(def ^:private conversion-needed? #{:tuple :vector :map :map-of})
+  **On sequences**
+  `:tuple`, `:vector` and `:sequential` all get transformed into CLJS vectors. When converting back to JS,
+  they are JS arrays."
+  (mtx/transformer
+    {:name :js
+     :decoders {:keyword           keyword
+                'keyword?          keyword
+                :qualified-keyword keyword
+                :vector            {:enter vec}
+                :sequential        {:enter vec}
+                :tuple             {:enter vec}
+                :map               {:enter decode-map}
+                :map-of            {:compile (fn [schema _]
+                                               (let [[key-s val-s] (mc/children schema)
+                                                     keydec (mc/decoder key-s js-transformer)
+                                                     valdec (mc/decoder val-s js-transformer)]
+                                                 #(decode-map-of keydec valdec %)))}
+                }
+     :encoders {:keyword           name
+                'keyword?          name
+                :qualified-keyword #(str (namespace %) "/" (name %))
+                :vector            {:leave clj->js}
+                :sequential        {:leave clj->js}
+                :tuple             {:leave clj->js}
+                :map               {:compile
+                                    (fn [schema _]
+                                      (let [js-props (into {} (for [[k props] (mc/children schema)
+                                                                    :when (:js/prop props)]
+                                                                [k (:js/prop props)]))
+                                            keyenc   (fn [k] (or (get js-props k)
+                                                                 (csk/->snake_case_string k)))]
+                                        {:leave #(encode-map % keyenc)}))}
+                :map-of            {:leave #(encode-map % name)}}}))
 
-(defmethod incoming :map-of [schema]
-  (let [vf          (incoming (:value schema))
-        keywordize? (-> schema :key :type (#{:keyword 'keyword?}))
-        recursive?  (-> schema :value :type conversion-needed?)
-        transform   (when (and (not (opaque? (:value schema)))
-                               (not= identity vf))
-                      vf)
-        opts        (cond-> {:recursive       (boolean recursive?)
-                             :keywordize-keys (boolean keywordize?)}
-                      keywordize? (assoc :prop->key csk/->kebab-case-keyword
-                                         :key->prop csk/->camelCaseString)
-                      transform   (assoc :transform transform))]
-    #(bean/bean % opts)))
+(defn incoming
+  "Returns a function for converting a JS value into CLJS data structures, based on a schema."
+  [schema]
+  (mc/decoder schema js-transformer))
```

Comment:
That feels like a dev-mode thing to do during tests, not so much in production.

WDYT?

Reply:
Can you help me see the potential harm in checking the shape in prod? I'm not a huge fan of decode{r} because it will pass through data that is completely invalid, which can be surprising. Maybe that is a low risk in prod. But yeah, we have a schema, and a shape that should match, so _at least in dev_ it's better to validate the decoded value here.

Reply:
I'm concerned about one thing for doing this in prod, and another for doing it anywhere.

In prod I'm concerned about performance. During the transition, when parts of metabase-lib are CLJC and parts TS, these objects get juggled back and forth pretty freely. That said, we can turn it on and see if it (a) makes any impact and (b) that `de/encoder`s are substantially faster than `coercer`s.

In anywhere, I'm concerned about the validation/treatment of `:any` - there are some places where that is intended to be a completely untouched JS object, that should be `identical?` when it comes back out, even after many `assoc-in` changes to sibling values. If I understand correctly, that is how `coercer` will treat them.

Also I recall that the FE unit tests use a release build of the CLJS  which makes the dev vs. prod point mostly moot. (This is not ideal but there's a Google Closure Library error if you use a debug one; to be fixed later.)

Reply:
> we can turn it on and see if it (a) makes any impact and (b) that de/encoders are substantially faster than coercers.

I think it makes sense. You're right that we should remove it if there are any performance issues.

Then as for `:any`, running the validation step shouldn't change the value, so it should behave the same way.

Reply:
I'm backing off on this and making it a TODO. Too many tests pass data that's missing fields. A worthwhile thing to fix later, but not something that should gate this early PR.